### PR TITLE
Add symmetric encryption and MAC support

### PR DESCRIPTION
**Pull request by danrr**
To enable stored messages to be protected using symmetric key encryption and validated
using message authentication codes, this set of changes adds support for storing
symmetric key material as Secret Key Packets, symmetric key encrypted session keys as
Public Key Encrypted Session Key Packets, and MAC tags as Signature Packets.

### CODE REVIEW

File: src/packet/public_key_encrypted_session_key.js [hunk 1]
```diff
@@ -106,7 +106,7 @@ class PublicKeyEncryptedSessionKeyPacket {
     ]);
     const algo = enums.write(enums.publicKey, this.publicKeyAlgorithm);
     this.encrypted = await crypto.publicKeyEncrypt(
-      algo, key.publicParams, data, key.getFingerprintBytes());
+      algo, key.publicParams, key.privateParams, data, key.getFingerprintBytes());
```
Comment (twiss) [inline src/packet/public_key_encrypted_session_key.js]: Instead of always passing the private params here, I would rather conditionally pass the private params instead of the public params if it's a symmetric algorithm. I know it's a little bit hacky, but that way there's fewer lines of code changed.
And similarly for `signature.verify` below.

File: src/crypto/crypto.js [hunk 2]
```diff
@@ -67,7 +75,24 @@ export async function publicKeyEncrypt(algo, publicParams, data, fingerprint) {
...
+      const { keyMaterial } = privateParams;
+      const aeadMode = enums.read(enums.aead, config.aeadMode);
+      const mode = crypto[aeadMode];
+      const { tagLength, ivLength } = mode;
+      const iv = await getRandomBytes(ivLength);
+      const modeInstance = await mode(algoName, keyMaterial);
+      const ciphertext = await modeInstance.encrypt(data, iv, new Uint8Array());
+      const c = ciphertext.subarray(0, -tagLength);
+      const t = ciphertext.subarray(-tagLength);
+      return { aeadMode: new AEADEnum(aeadMode), iv, c: new ByteArray(c), t };
```
Comment (twiss) [inline src/crypto/crypto.js]: I wouldn't encode the tag separately (i.e. I'd just keep it included in the ciphertext, the only thing it changes is the encoded length before, and I think it makes things simpler)

File: src/packet/public_key_encrypted_session_key.js (hunk 1 above)
Comment (danrr) [reply src/packet/public_key_encrypted_session_key.js]: You need both, since the public params hold the symmetric cipher.

Comment (twiss) [reply src/packet/public_key_encrypted_session_key.js]: Ah yeah, alrighty.

File: src/crypto/crypto.js (hunk 2 above)
Comment (danrr) [reply src/crypto/crypto.js]: The spec lists ciphertext and tag separately for AEAD packets (tag 20), I was going off that when writing out the draft RFC and this lines up with the draft. The symmetric encryption draft went from

<ul> <li>A symmetric key encryption of "m" dependent on cipher and AEAD mode</li>
                <li>An authentication tag generated by the AEAD mode."</li></ul>
to:
<ul> <li>A symmetric key encryption of "m" dependent on cipher and AEAD mode prefixed with a one-octet length </li>
                <li>An authentication tag generated by the AEAD mode."</li></ul>

It would now have to be something like:

<ul> <li>A symmetric key encryption of "m" dependent on cipher and AEAD mode concatenated with the authentication tag generated by the AEAD mode,  prefixed with a one-octet length of the concatenation</li>
              </ul>
It's a bit wordy :D.

Comment (danrr) [reply src/crypto/crypto.js]: I think having the two separate is conceptually easier to understand and it aligns with the existing stuff in the spec.

Comment (twiss) [reply src/crypto/crypto.js]: Are you referring to
```
   o  Encrypted data, the output of the selected symmetric-key cipher
      operating in the given AEAD mode.

   o  A final, summary authentication tag for the AEAD mode.
```
? Because these are two separate things. The first bullet point includes both the ciphertext and authentication tag of each chunk; the second bullet point is again a full invocation of the AEAD construct but on an empty plaintext, which means you only get an authentication tag.

Then in the text below it does mention the ciphertext and authentication tag separately, but not in separate bullet points, and the spec currently never asks you to split them up in code (they are always stored immediately next to each other and not counted separately). So I think we shouldn't do so here, either.
